File: app/routes/catalog/schemas.py
from typing import Optional, List, Dict
import uuid
from datetime import datetime

from pydantic import BaseModel, field_validator
from sqlalchemy.orm import joinedload, selectinload
from sqlalchemy.orm.strategy_options import _AbstractLoad

from app.routes.utils import ORMModel
from core.models.catalog import Product, SKU, ProductVariant
from core.services.schemas.schema import ProductType


class PrintingResponseSchema(ORMModel):
    id: uuid.UUID
    name: str


class LanguageResponseSchema(ORMModel):
    id: uuid.UUID
    name: str
    abbreviation: str


class ConditionResponseSchema(ORMModel):
    id: uuid.UUID
    name: str
    abbreviation: str


class ProductBaseResponseSchema(ORMModel):
    id: uuid.UUID
    name: str
    tcgplayer_url: str
    image_url: str
    product_type: ProductType
    data: List[Dict[str, str]]
    rarity: Optional[str] = None
    number: Optional[str] = None


class SetBaseResponseSchema(ORMModel):
    id: uuid.UUID
    name: str
    code: str
    release_date: datetime


class SKUBaseResponseSchema(ORMModel):
    id: uuid.UUID
    condition: ConditionResponseSchema
    printing: PrintingResponseSchema
    language: LanguageResponseSchema

    @classmethod
    def get_load_options(cls) -> list[_AbstractLoad]:
        return [
            joinedload(SKU.condition),
            joinedload(SKU.printing),
            joinedload(SKU.language),
        ]


class ProductVariantResponseSchema(ORMModel):
    id: uuid.UUID
    product: ProductBaseResponseSchema
    set: SetBaseResponseSchema
    printing: PrintingResponseSchema

    @classmethod
    def get_load_options(cls) -> list[_AbstractLoad]:
        return [
            joinedload(ProductVariant.product).options(
                joinedload(Product.set),
            ),
            joinedload(ProductVariant.printing),
        ]


class ProductWithSetAndSKUsResponseSchema(ProductBaseResponseSchema):
    set: SetBaseResponseSchema
    skus: list[SKUBaseResponseSchema]

    @classmethod
    def get_load_options(cls) -> list[_AbstractLoad]:
        return [
            selectinload(Product.skus).options(
                *SKUBaseResponseSchema.get_load_options(),
            ),
            joinedload(Product.set),
        ]

    @field_validator("skus", mode="before")
    def sort_skus(
        cls, skus: list[SKUBaseResponseSchema]
    ) -> list[SKUBaseResponseSchema]:
        """
        Sort the list of SKUs first by the condition's name and then by the printing's name.
        Adjust the lambda key as needed if the fields to sort by differ.
        """
        return sorted(skus, key=lambda sku: (sku.condition.id, sku.printing.name))


class SKUWithProductResponseSchema(SKUBaseResponseSchema):
    product: ProductWithSetAndSKUsResponseSchema

    @classmethod
    def get_load_options(cls) -> list[_AbstractLoad]:
        return [
            selectinload(SKU.product).options(
                *ProductWithSetAndSKUsResponseSchema.get_load_options()
            )
        ]


class ProductSearchResultSchema(ProductBaseResponseSchema):
    """Deprecated: use ProductVariantResponseSchema results instead."""

    set: SetBaseResponseSchema
    variants: list[ProductVariantResponseSchema]

    @classmethod
    def get_load_options(cls) -> list[_AbstractLoad]:
        return [
            joinedload(Product.set),
            selectinload(Product.variants).options(
                *ProductVariantResponseSchema.get_load_options()
            ),
        ]


class ProductSearchResponseSchema(BaseModel):
    results: list[ProductVariantResponseSchema]


class CatalogResponseSchema(ORMModel):
    id: uuid.UUID
    display_name: str


class CatalogsResponseSchema(BaseModel):
    catalogs: list[CatalogResponseSchema]


class ProductTypesResponseSchema(BaseModel):
    product_types: list[ProductType]


class SetsResponseSchema(BaseModel):
    sets: list[SetBaseResponseSchema]


class TopPricedCardSchema(BaseModel):
    sku_id: uuid.UUID
    product_name: str
    condition: str
    printing: str
    language: str
    price: float


class HistoricalPriceComparisonSchema(BaseModel):
    current_total_market_value: float
    historical_total_market_value: float | None
    growth_percentage: float | None
    current_top_priced_card: TopPricedCardSchema | None
    historical_top_priced_card: TopPricedCardSchema | None
    top_card_growth_percentage: float | None


class ProductSearchRequestParams(BaseModel):
    query: str
    catalog_id: Optional[uuid.UUID] = None
    product_type: Optional[ProductType] = None
    set_id: Optional[uuid.UUID] = None
    limit: Optional[int] = None

    class Config:
        extra = "forbid"

File: web/app/catalog/schemas.ts
import { z } from "zod";

const UUIDSchema = z.string().uuid("Expected a UUID string");

export const ProductTypeSchema = z.enum(["CARDS", "SEALED"]);

export const PrintingResponseSchema = z.object({
  id: UUIDSchema,
  name: z.string(),
});

export const LanguageResponseSchema = z.object({
  id: UUIDSchema,
  name: z.string(),
  abbreviation: z.string(),
});

export const ConditionResponseSchema = z.object({
  id: UUIDSchema,
  name: z.string(),
  abbreviation: z.string(),
});

export const SetBaseResponseSchema = z.object({
  id: UUIDSchema,
  name: z.string(),
  code: z.string(),
  release_date: z.string().datetime(),
});

export const SKUBaseResponseSchema = z.object({
  id: UUIDSchema,
  condition: ConditionResponseSchema,
  printing: PrintingResponseSchema,
  language: LanguageResponseSchema,
  lowest_listing_price_total: z.number(),
});

export const ProductBaseResponseSchema = z.object({
  id: UUIDSchema,
  name: z.string(),
  tcgplayer_url: z.string(),
  image_url: z.string(),
  product_type: ProductTypeSchema,
  data: z.array(z.record(z.string(), z.string())),
  rarity: z.string().nullable(),
  number: z.string().nullable(),
});

export const ProductWithSetAndSKUsResponseSchema =
  ProductBaseResponseSchema.extend({
    set: SetBaseResponseSchema,
    skus: z.array(SKUBaseResponseSchema),
  });

export const SKUWithProductResponseSchema = SKUBaseResponseSchema.extend({
  product: ProductWithSetAndSKUsResponseSchema,
});

export const ProductSearchResponseSchema = z.object({
  results: z.array(ProductWithSetAndSKUsResponseSchema),
  total: z.number(),
  page: z.number(),
  limit: z.number(),
  has_next: z.boolean(),
  has_prev: z.boolean(),
});

export type ProductType = z.infer<typeof ProductTypeSchema>;
export type PrintingResponse = z.infer<typeof PrintingResponseSchema>;
export type LanguageResponse = z.infer<typeof LanguageResponseSchema>;
export type ConditionResponse = z.infer<typeof ConditionResponseSchema>;
export type SetBaseResponse = z.infer<typeof SetBaseResponseSchema>;
export type SKUBaseResponse = z.infer<typeof SKUBaseResponseSchema>;
export type ProductBaseResponse = z.infer<typeof ProductBaseResponseSchema>;
export type ProductWithSetAndSKUsResponse = z.infer<
  typeof ProductWithSetAndSKUsResponseSchema
>;
export type SKUWithProductResponse = z.infer<
  typeof SKUWithProductResponseSchema
>;
export type ProductSearchResponse = z.infer<typeof ProductSearchResponseSchema>;

export type { SKUMarketDataItem } from "@/features/market/types";
